### PR TITLE
lib/scanner: Use fs.Filesystem for all operations

### PR DESCRIPTION
**Pull request by calmh**
### Purpose

One more step on the path of the great refactoring. Touches rwfolder a little bit since it uses the Lstat from fs as well, but mostly this is just on the scanner as rwfolder is scheduled for a later refactor.

There are a couple of usages of fs.DefaultFilesystem that will in the end become a filesystem injected from the top, but that comes later.

### Testing

Much of this is still unit tested, and I ran a couple of manual integration tests on top.

### CODE REVIEW

File: lib/fs/basicfs.go [hunk 1]
```diff
@@ -71,11 +71,32 @@ func (f *BasicFilesystem) DirNames(name string) ([]string, error) {
...
+}
+
+// fsFile implements the fs.File interface on top of an os.File
+type fsFile struct {
+	*os.File
+}
+
+func (f fsFile) Stat() (FileInfo, error) {
+	info, err := f.File.Stat()
+	if err != nil {
```
Comment (AudriusButkevicius) [inline lib/fs/basicfs.go]: Doesn't this need the magic stat treatment?

Comment (AudriusButkevicius) [inline lib/fs/basicfs.go]: I guess you could pass a handle to the underlying fs, and it could do f.fs.Stat()

File: lib/fs/mtimefs_test.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ func TestMtimeFS(t *testing.T) {
 	// a random time with nanosecond precision
 	testTime := time.Unix(1234567890, 123456789)
 
-	mtimefs := NewMtimeFS(make(mapStore))
+	mtimefs := NewMtimeFS(make(mapStore), DefaultFilesystem)
```
Comment (AudriusButkevicius) [inline lib/fs/mtimefs_test.go]: Can we swap the args around given mtimefs needs a backing fs.

File: lib/model/rwfolder.go [hunk 3]
```diff
@@ -678,7 +678,7 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo) {
 	// It's OK to change mode bits on stuff within non-writable directories.
 	if f.ignorePermissions(file) {
 		f.dbUpdates <- dbUpdateJob{file, dbUpdateHandleDir}
-	} else if err := os.Chmod(realName, mode|(info.Mode()&retainBits)); err == nil {
+	} else if err := os.Chmod(realName, mode|(os.FileMode(info.Mode())&retainBits)); err == nil {
 		f.dbUpdates <- dbUpdateJob{file, dbUpdateHandleDir}
```
Comment (AudriusButkevicius) [inline lib/model/rwfolder.go]: On the phone so ignore my ignorance. This is just an int, so why do we need to change the type?

File: lib/fs/basicfs.go (hunk 1 above)
Comment (calmh) [inline lib/fs/basicfs.go]: I don't think so. We haven't treated these specifically before, just done Stat() on the os-returned fd.

That said, I was considering removing the android-specialcasing anyway to see if it's still broken or not... But opted to bring it along in the end.

File: lib/model/rwfolder.go (hunk 3 above)
Comment (calmh) [inline lib/model/rwfolder.go]: Chmod wants an os.FileMode, and we return an fs.FileMode. The others are just numeric constants, so in the end the expression becomes fs.FileMode unless converted.

File: lib/scanner/walk.go [hunk 4]
```diff
@@ -80,18 +80,14 @@ type CurrentFiler interface {
...
 func Walk(cfg Config) (chan protocol.FileInfo, error) {
 	w := walker{cfg}
 
 	if w.CurrentFiler == nil {
 		w.CurrentFiler = noCurrentFiler{}
 	}
-	if w.Lstater == nil {
-		w.Lstater = defaultLstater{}
+	if w.Filesystem == nil {
+		w.Filesystem = fs.DefaultFilesystem
```
Comment (calmh) [inline lib/scanner/walk.go]: Not by the tests. It seems sane to have a fallback, but in the end I don't feel strongly about it either way.

File: lib/fs/basicfs.go (hunk 1 above)
Comment (AudriusButkevicius) [inline lib/fs/basicfs.go]: We should definately keep it, unless stdlib addresses it.

File: lib/model/rwfolder.go (hunk 3 above)
Comment (AudriusButkevicius) [inline lib/model/rwfolder.go]: I guess this will become a fs.Chmod at which point these will go away. Fine.

File: lib/scanner/walk.go (hunk 4 above)
Comment (AudriusButkevicius) [inline lib/scanner/walk.go]: Ok let's keep it.

File: lib/fs/basicfs.go (hunk 1 above)
Comment (calmh) [inline lib/fs/basicfs.go]: Yep. It is kept. The code above is equal to what we had before.

File: lib/scanner/blockqueue.go [hunk 5]
```diff
@@ -42,7 +42,7 @@ func newParallelHasher(dir string, blockSize, workers int, outbox, inbox chan pr
 
 // HashFile hashes the files and returns a list of blocks representing the file.
 func HashFile(path string, blockSize int, counter Counter, useWeakHashes bool) ([]protocol.BlockInfo, error) {
-	fd, err := os.Open(path)
+	fd, err := fs.DefaultFilesystem.Open(path)
```
Comment (AudriusButkevicius) [inline lib/scanner/blockqueue.go]: Hmm, this should probably bubble down all the way from the scanner if you want to declare that "scanner conversion is finished".

Comment (calmh) [inline lib/scanner/blockqueue.go]: Mmmyeah that's a bit lazy, let me refactor that walker thing.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

File: lib/fs/filesystem.go [hunk 6]
```diff
@@ -57,12 +59,20 @@ type FileInfo interface {
 // FileMode is similar to os.FileMode
 type FileMode uint32
 
+// ModePerm is the equivalent of os.ModePerm
+const ModePerm = FileMode(os.ModePerm)
+
 // DefaultFilesystem is the fallback to use when nothing explicitly has
 // been passed.
 var DefaultFilesystem Filesystem = new(BasicFilesystem)
```
Comment (imsodin) [inline lib/fs/filesystem.go]: Not directly related to this PR, but why isn't this ` = NewBasicFilesystem()`?

Comment (calmh) [inline lib/fs/filesystem.go]: Probably I introduced that method later and forgot. Good catch, fixed!

**Comment by imsodin** [comment]
@st-review lgtm

**Comment by st-review** [comment]
:ok_hand: Merged as 4253f22680dfc700759529c553554a106fa16a6a. Thanks, @calmh!

**Pull request closed without merging**